include/dice/template-library/limit_allocator.hpp

```diff
@@ -0,0 +1,174 @@
+#ifndef DICE_TEMPLATELIBRARY_LIMITALLOCATOR_HPP
+#define DICE_TEMPLATELIBRARY_LIMITALLOCATOR_HPP
+
+#include <atomic>
+#include <cstddef>
+#include <memory>
+#include <new>
+#include <type_traits>
+#include <utility>
+
+namespace dice::template_library {
+
+	/**
+	 * The synchronization policy of a limit_allocator
+	 */
+	enum struct limit_allocator_syncness : bool {
+		sync, ///< thread-safe (synchronized)
+		unsync ///< not thread-safe (unsynchronized)
+	};
+
+	namespace detail_limit_allocator {
+		template<limit_allocator_syncness syncness>
+		struct limit_allocator_control_block;
+
+		template<>
+		struct limit_allocator_control_block<limit_allocator_syncness::sync> {
+			std::atomic<size_t> bytes_left = 0;
+
+			void allocate(size_t n_bytes) {
+				auto old = bytes_left.load(std::memory_order_relaxed);
+
+				do {
+					if (old < n_bytes) [[unlikely]] {
+						throw std::bad_alloc{};
+					}
+				} while (!bytes_left.compare_exchange_weak(old, old - n_bytes, std::memory_order_relaxed, std::memory_order_relaxed));
+			}
+
+			void deallocate(size_t n_bytes) noexcept {
+				bytes_left.fetch_add(n_bytes, std::memory_order_relaxed);
+			}
+		};
+
+		template<>
+		struct limit_allocator_control_block<limit_allocator_syncness::unsync> {
+			size_t bytes_left = 0;
+
+			void allocate(size_t n_bytes) {
+				if (bytes_left < n_bytes) [[unlikely]] {
+					throw std::bad_alloc{};
+				}
+				bytes_left -= n_bytes;
+			}
+
+			void deallocate(size_t n_bytes) noexcept {
+				bytes_left += n_bytes;
+			}
+		};
+	} // namespace detail_limit_allocator
+
+	/**
+	 * Allocator wrapper that limits the amount of memory its underlying allocator
+	 * is allowed to allocate.
+	 *
+	 * @tparam T value type of the allocator (the thing that it allocates)
+	 * @tparam Allocator the underlying allocator
+	 * @tparam syncness determines the synchronization of the limit
+	 */
+    template<typename T, template<typename> typename Allocator = std::allocator, limit_allocator_syncness syncness = limit_allocator_syncness::sync>
+	struct limit_allocator {
+    	using control_block_type = detail_limit_allocator::limit_allocator_control_block<syncness>;
+		using value_type = T;
+    	using upstream_allocator_type = Allocator<T>;
```

Comment:
Isn't it "downstream" not "upstream"?

Reply:
I guess it depends on how you look at it, I chose the same naming as the std: https://en.cppreference.com/w/cpp/memory/synchronized_pool_resource/upstream_resource

Reply:
true. then let's keep upstream